src/config.rs

```diff
@@ -0,0 +1,350 @@
+                Some("K") => Some(1024 * size),
+                Some("M") => Some(1024 * 1024 * size),
+                Some("G") => Some(1024 * 1024 * 1024 * size),
+                Some("T") => Some(1024 * 1024 * 1024 * 1024 * size),
+                _ => None,
+            }
+        })
+}
+
+#[derive(Debug, PartialEq)]
+#[derive(Deserialize)]
+pub struct AzureCacheConfig;
+
+#[derive(Debug, PartialEq)]
+#[derive(Deserialize)]
+pub struct DiskCacheConfig {
+    pub dir: PathBuf,
+    // TODO: use deserialize_with to allow human-readable sizes in toml
+    pub size: u64,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+#[derive(Deserialize)]
+pub enum GCSCacheRWMode {
+    #[serde(rename = "READ_ONLY")]
+    ReadOnly,
+    #[serde(rename = "READ_WRITE")]
+    ReadWrite,
+}
+
+#[derive(Debug, PartialEq)]
+#[derive(Deserialize)]
+pub struct GCSCacheConfig {
+    pub bucket: String,
+    pub cred_path: Option<PathBuf>,
+    pub rw_mode: GCSCacheRWMode,
+}
+
+#[derive(Debug, PartialEq)]
+#[derive(Deserialize)]
+pub struct MemcachedCacheConfig {
+    pub url: String,
+}
+
+#[derive(Debug, PartialEq)]
+#[derive(Deserialize)]
+pub struct RedisCacheConfig {
+    pub url: String,
+}
+
+#[derive(Debug, PartialEq)]
+#[derive(Deserialize)]
+pub struct S3CacheConfig {
+    pub bucket: String,
+    pub endpoint: String,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum CacheType {
+    Azure(AzureCacheConfig),
+    Disk(DiskCacheConfig),
+    GCS(GCSCacheConfig),
+    Memcached(MemcachedCacheConfig),
+    Redis(RedisCacheConfig),
+    S3(S3CacheConfig),
+}
+
+#[derive(Debug, Default)]
+#[derive(Deserialize)]
+pub struct CacheConfigs {
+    azure: Option<AzureCacheConfig>,
+    disk: Option<DiskCacheConfig>,
+    gcs: Option<GCSCacheConfig>,
+    memcached: Option<MemcachedCacheConfig>,
+    redis: Option<RedisCacheConfig>,
+    s3: Option<S3CacheConfig>,
+}
+
+impl CacheConfigs {
+    /// Return a vec of the available cache types in a specific
+    /// consistent ordering
+    fn into_vec(self) -> Vec<CacheType> {
+        let CacheConfigs {
+            azure, disk, gcs, memcached, redis, s3
+        } = self;
+
+        s3.map(CacheType::S3).into_iter()
+            .chain(redis.map(CacheType::Redis))
+            .chain(memcached.map(CacheType::Memcached))
+            .chain(gcs.map(CacheType::GCS))
+            .chain(azure.map(CacheType::Azure))
+            .chain(disk.map(CacheType::Disk))
+            .collect()
+    }
+}
+
+#[derive(Debug, Default)]
+#[derive(Deserialize)]
+pub struct FileConfig {
+    cache: CacheConfigs,
+}
+
+fn try_read_config_file(path: &Path) -> Option<FileConfig> {
+    let mut file = File::open(path)
+        .map_err(|e| debug!("Couldn't open config file: {}", e))
+        .ok()?;
+
+    let mut string = String::new();
+    file.read_to_string(&mut string)
+        .map_err(|e| warn!("Failed to read config file: {}", e))
+        .ok()?;
+
+    let toml: toml::Value = toml::from_str(&string)
```

Comment:
You could just make this `let cfg: FileConfig = toml::from_str(&string)...` and skip a step, right?